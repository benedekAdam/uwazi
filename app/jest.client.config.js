module.exports = {
  name: 'client',
  displayName: 'Client',
  testMatch: [
    '**/react/**/specs/*spec.js?(x)'
  ],
  testPathIgnorePatterns: [],
  setupTestFrameworkScriptFile: '<rootDir>/setUpJestClient.js',
  moduleNameMapper: {
    '\\.(css|scss)$': 'identity-obj-proxy'
  },
  snapshotSerializers: ['enzyme-to-json/serializer']
};
