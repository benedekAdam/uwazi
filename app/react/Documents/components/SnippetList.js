/* eslint-disable react/no-danger */
/* eslint-disable react/no-array-index-key */
import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { t, I18NLink } from 'app/I18N';

function getFieldLabel(field, template) {
  if (field === 'title') {
    return t('System', 'Title');
  }
  if (field.startsWith('metadata.') && template) {
    const name = field.split('.')[1];
    const property = template.get('properties').find(p => p.get('name') === name);
    if (property) {
      return t(template.get('_id'), property.get('label'));
    }
  }
  return field;
}

export const MetadataFieldSnippets = ({ fieldSnippets, documentViewUrl, template }) => (
  <React.Fragment>
    <li className="snippet-list-item-header metadata-snippet-header">
      <I18NLink to={documentViewUrl}>
        { getFieldLabel(fieldSnippets.get('field'), template) }
      </I18NLink>
    </li>
    {fieldSnippets.get('texts').map((snippet, index) => (
      <li key={index} className="snippet-list-item metadata-snippet">
        <span dangerouslySetInnerHTML={{ __html: snippet }} />
      </li>
    ))}
  </React.Fragment>
);

MetadataFieldSnippets.propTypes = {
  fieldSnippets: PropTypes.shape({
    texts: PropTypes.array,
    field: PropTypes.string
  }).isRequired,
  documentViewUrl: PropTypes.string.isRequired,
  template: PropTypes.shape({
    get: PropTypes.func
  })
};

MetadataFieldSnippets.defaultProps = {
  template: undefined
};

export const DocumentContentSnippets = ({ scrollToPage, documentSnippets, documentViewUrl, searchTerm }) => (
  <React.Fragment>
    <li className="snippet-list-item-header fulltext-snippet-header">
      {t('System', 'Document contents')}
    </li>
    {documentSnippets.map((snippet, index) => (
      <li key={index} className="snippet-list-item fulltext-snippet">
        <I18NLink
          onClick={() => scrollToPage(snippet.get('page'))}
          to={`${documentViewUrl}?page=${snippet.get('page')}&searchTerm=${searchTerm || ''}`}
        >
          {snippet.get('page')}
        </I18NLink>
        <span dangerouslySetInnerHTML={{ __html: snippet.get('text') }} />
      </li>
    ))}
  </React.Fragment>
);

DocumentContentSnippets.propTypes = {
  scrollToPage: PropTypes.func.isRequired,
  documentSnippets: PropTypes.shape({
    map: PropTypes.func
  }).isRequired,
  documentViewUrl: PropTypes.string.isRequired,
  searchTerm: PropTypes.string.isRequired
};

export const SnippetList = ({ snippets, documentViewUrl, searchTerm, scrollToPage, template }) => (
  <ul className="snippet-list">
    {snippets.get('metadata').map(fieldSnippets => (
      <MetadataFieldSnippets
        key={fieldSnippets.get('field')}
        fieldSnippets={fieldSnippets}
        template={template}
        documentViewUrl={documentViewUrl}
      />
    ))}
    {snippets.get('fullText').size ? (
      <DocumentContentSnippets
        documentSnippets={snippets.get('fullText')}
        documentViewUrl={documentViewUrl}
        scrollToPage={scrollToPage}
        searchTerm={searchTerm}
      />
     ) : ''}
  </ul>
);

SnippetList.propTypes = {
  doc: PropTypes.shape({
    get: PropTypes.func
  }).isRequired,
  documentViewUrl: PropTypes.string.isRequired,
  scrollToPage: PropTypes.func.isRequired,
  searchTerm: PropTypes.string.isRequired,
  snippets: PropTypes.shape({
    get: PropTypes.func
  }).isRequired,
  template: PropTypes.shape({
    get: PropTypes.func
  })
};

SnippetList.defaultProps = {
  template: undefined
};

export const mapStateToProps = (state, ownProps) => ({
  template: state.templates.find(tmpl => tmpl.get('_id') === ownProps.doc.get('template'))
});

export default connect(mapStateToProps)(SnippetList);
