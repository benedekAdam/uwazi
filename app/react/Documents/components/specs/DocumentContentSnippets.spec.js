import React from 'react';
import { shallow } from 'enzyme';

import Immutable from 'immutable';
import { I18NLink } from 'app/I18N';
import { browserHistory } from 'react-router';
import { DocumentContentSnippets } from '../SnippetList';

describe('SnippetList', () => {
  let component;
  let props;

  const render = () => {
    component = shallow(<DocumentContentSnippets {...props}/>);
  };

  beforeEach(() => {
    spyOn(browserHistory, 'getCurrentLocation').and.returnValue({ pathname: 'path', query: { page: 1 } });
    props = {
      doc: Immutable.fromJS({ _id: 'id', sharedId: 'sharedId', type: 'document' }),
      searchTerm: 'snippet',
      documentViewUrl: '/document/sharedId',
      scrollToPage: jest.fn(),
      documentSnippets: Immutable.fromJS([
        { text: 'first <b>snippet 1</b> found', page: 1 },
        { text: 'second <b>snippet 3</b> found', page: 2 },
        { text: 'third <b>snippet 3</b> found', page: 3 }
      ]),
    };
  });

  it('should render all document snippets', () => {
    render();
    const snippets = component;
    expect(snippets).toMatchSnapshot();
  });

  it('should scrollToPage when click on a snippet link', () => {
    props.doc = Immutable.fromJS({ _id: 'id', sharedId: 'sharedId', type: 'document' });
    props.scrollToPage = jest.fn();
    render();
    component.find(I18NLink).at(1).simulate('click');
    expect(props.scrollToPage).toHaveBeenCalledWith(2);
  });
});
