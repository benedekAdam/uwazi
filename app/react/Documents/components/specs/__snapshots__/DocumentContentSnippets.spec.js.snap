// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SnippetList should render all document snippets 1`] = `
<React.Fragment>
  <li
    className="snippet-list-item-header fulltext-snippet-header"
  >
    Document contents
  </li>
  <li
    className="snippet-list-item fulltext-snippet"
    key="0"
  >
    <Connect(I18NLink)
      onClick={[Function]}
      to="/document/sharedId?page=1&searchTerm=snippet"
    >
      1
    </Connect(I18NLink)>
    <span
      dangerouslySetInnerHTML={
        Object {
          "__html": "first <b>snippet 1</b> found",
        }
      }
    />
  </li>
  <li
    className="snippet-list-item fulltext-snippet"
    key="1"
  >
    <Connect(I18NLink)
      onClick={[Function]}
      to="/document/sharedId?page=2&searchTerm=snippet"
    >
      2
    </Connect(I18NLink)>
    <span
      dangerouslySetInnerHTML={
        Object {
          "__html": "second <b>snippet 3</b> found",
        }
      }
    />
  </li>
  <li
    className="snippet-list-item fulltext-snippet"
    key="2"
  >
    <Connect(I18NLink)
      onClick={[Function]}
      to="/document/sharedId?page=3&searchTerm=snippet"
    >
      3
    </Connect(I18NLink)>
    <span
      dangerouslySetInnerHTML={
        Object {
          "__html": "third <b>snippet 3</b> found",
        }
      }
    />
  </li>
</React.Fragment>
`;
